From dcd3253cb2091d21c42d65947cfd28b4fdd77896 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Thu, 14 Dec 2017 17:42:42 +0000
Subject: [PATCH] rtc-rs5c372: add oftree support for osc adjustment

        This commit moves osc adjustment configuration to
        oftree setup. By this we are able to set platform
        specific configuration.

fix build
---
 drivers/rtc/rtc-rs5c372.c | 38 ++++++++++++++++++++++++++++++--------
 1 file changed, 30 insertions(+), 8 deletions(-)

diff --git a/drivers/rtc/rtc-rs5c372.c b/drivers/rtc/rtc-rs5c372.c
index cf093a3..c926738 100644
--- a/drivers/rtc/rtc-rs5c372.c
+++ b/drivers/rtc/rtc-rs5c372.c
@@ -850,14 +850,36 @@ static int rs5c372_probe(struct i2c_client *client,
 	}
 
 #ifdef	NEED_TRIM
-	pr_debug("%s: set osc adjustment to %s", __func__, "0x0d (13)");
-	rs5c372_set_trim(client, 0, 0, 13); /* Calculation is based on an empirical drift value of +32.47 ppm.
-					     * Set trim value to (32768.00 * 32.47 ppm * 10) + 1 = 11.6398 ~ 12.
-					     * Measurements provide better values for an adjustment value of 13
-					     * which results in a drift value of -1.41 ppm. Also Hardware measurements
-					     * of osc frequency indicate the need of a higher value. So we set 13 here.
-					     * See documentation of r2221 on page 34.
-					     */
+        {
+                struct device_node *node = client->dev.of_node;
+                u8 dev_bit, dec;
+                int trim;
+                u32 trim_data[3];
+
+                if (node) {
+                        int ret = of_property_read_u32_array(node, "trim-data", trim_data,
+                                                ARRAY_SIZE(trim_data));
+                        if (!ret) {
+                                dev_bit = (u8) trim_data[0];
+                                dec = (u8) trim_data[1];
+                                trim = (int) trim_data[2];
+
+                                dev_info(&client->dev, "set osc adjustment to 0x%x (%d)\n",
+                                                trim, trim);
+
+                                /* Calculation is based on an empirical drift value of +32.47 ppm.
+                                 * Set trim value to (32768.00 * 32.47 ppm * 10) + 1 = 11.6398 ~ 12.
+                                 * Measurements provide better values for an adjustment value of 13
+                                 * which results in a drift value of -1.41 ppm. Also Hardware measurements
+                                 * of osc frequency indicate the need of a higher value. So we set 13 here.
+                                 * See documentation of r2221 on page 34.
+                                 */
+                                rs5c372_set_trim(client, dev_bit, dec, trim);
+                        } else {
+                                dev_info(&client->dev, "skip osc adjustment\n");
+                        }
+                }
+        }
 #endif
 
 	dev_info(&client->dev, "%s found, %s\n",
-- 
2.7.4

